### PR TITLE
DN supports objectio to read and write data

### PR DESCRIPTION
**Pull request by LeftHandCold**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3532 #3919

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/vm/engine/tae/catalog/metabase.go [hunk 1]
```diff
@@ -60,7 +60,37 @@ func (be *MetaBaseEntry) PPString(level common.PPLevel, depth int, prefix string
...
+func (be *MetaBaseEntry) GetMetaLoc() string {
+	be.RLock()
+	defer be.RUnlock()
+	if be.GetNodeLocked() == nil {
+		return ""
+	}
+	str := be.GetNodeLocked().(*MetadataMVCCNode).MetaLoc
+	return str
+}
+func (be *MetaBaseEntry) GetDeltaLoc() string {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/catalog/metabase.go:72]: Currently, there may be no problem here. But later we should always get an mvcc-value.

File: pkg/vm/engine/tae/catalog/metabase.go [hunk 2]
```diff
@@ -121,10 +151,32 @@ func (be *MetaBaseEntry) UpdateAttr(txn txnif.TxnReader, node *MetadataMVCCNode)
...
+	entry.UpdateMetaLoc(metaloc)
+	return
+}
+
+func (be *MetaBaseEntry) UpdateDeltaLoc(txn txnif.TxnReader, deltaloc string) (isNewNode bool, err error) {
+	be.Lock()
+	defer be.Unlock()
+	needWait, txnToWait := be.NeedWaitCommitting(txn.GetStartTS())
+	if needWait {
+		be.RUnlock()
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/catalog/metabase.go:169]: This entry was Lock()'ed and now RUnlock()?

File: pkg/vm/engine/tae/catalog/metabase.go [hunk 3]
```diff
@@ -112,7 +142,7 @@ func (be *MetaBaseEntry) DeleteLocked(txn txnif.TxnReader) (isNewNode bool, err
 	return
 }
 
-func (be *MetaBaseEntry) UpdateAttr(txn txnif.TxnReader, node *MetadataMVCCNode) (isNewNode bool, err error) {
+func (be *MetaBaseEntry) UpdateMetaLoc(txn txnif.TxnReader, metaloc string) (isNewNode bool, err error) {
 	be.Lock()
 	defer be.Unlock()
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/catalog/metabase.go:147]: RUnlock a  Lock'ed?

File: pkg/vm/engine/tae/catalog/block.go [hunk 4]
```diff
@@ -147,8 +147,8 @@ func (entry *BlockEntry) InitData(factory DataFactory) {
 }
 func (entry *BlockEntry) GetBlockData() data.Block { return entry.blkData }
 func (entry *BlockEntry) GetSchema() *Schema       { return entry.GetSegment().GetTable().GetSchema() }
-func (entry *BlockEntry) GetFileTs() (types.TS, error) {
-	return entry.GetBlockData().GetBlockFile().ReadTS()
+func (entry *BlockEntry) GetFileTs() (ts types.TS, err error) {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/catalog/block.go:150]: GetFileTs?

File: pkg/objectio/fs.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+func TmpNewFileservice(dir string) fileservice.FileService {
+	c := fileservice.Config{
+		Name:    "LOCAL",
+		Backend: "DISK",
+		DataDir: dir,
+	}
+	service, err := fileservice.NewFileService(c)
+	if err != nil {
+		panic(any(fmt.Sprintf("NewFileService failed: %s", err.Error())))
```
Comment (fengttt) [inline pkg/objectio/fs.go:38]: Why don't you panci(err)?

Comment (LeftHandCold) [reply pkg/objectio/fs.go:38]: This is a temporary function, fileservice needs to be initialized in main.go and then passed in, so there is no need to create fileservice here

File: pkg/vm/engine/tae/tables/block.go [hunk 6]
```diff
@@ -156,7 +154,7 @@ func newBlock(meta *catalog.BlockEntry, segFile file.Segment, bufMgr base.INodeM
 	}
 	block.mvcc.SetMaxVisible(ts)
 	block.ckpTs.Store(ts)
-	if !ts.IsEmpty() {
+	if len(meta.GetMetaLoc()) > 0 {
```
Comment (XuPeng-SH) [inline pkg/vm/engine/tae/tables/block.go:157]: if $string != "" {
   xxxxxx
}

File: pkg/objectio/fs.go (hunk 5 above)
Comment (fengttt) [reply pkg/objectio/fs.go:35]: We should always panci a moerr.

File: pkg/vm/engine/tae/catalog/metabase.go (hunk 1 above)
Comment (jiangxinmeng1) [reply pkg/vm/engine/tae/catalog/metabase.go:72]: GetVisibleDeltaLoc can get mvcc-value by timestamp

File: pkg/vm/engine/tae/catalog/metabase.go [hunk 7]
```diff
@@ -60,7 +60,37 @@ func (be *MetaBaseEntry) PPString(level common.PPLevel, depth int, prefix string
 	s := fmt.Sprintf("%s%s%s", common.RepeatStr("\t", depth), prefix, be.StringLocked())
 	return s
 }
+func (be *MetaBaseEntry) GetMetaLoc() string {
```
[the remark this answers is not in the document]
Comment (jiangxinmeng1) [reply pkg/vm/engine/tae/catalog/metabase.go:63]: GetVisibleMetaLoc can get mvcc-value by timestamp

**Review by XuPeng-SH — APPROVED** [review]
Need some follow-up PR to continue to improve

**Comment by LeftHandCold** [comment]
@matrix-meow rerun failed

**Comment by LeftHandCold** [comment]
@matrix-meow rerun failed

**Comment by LeftHandCold** [comment]
@mertifyio refresh

**Comment by LeftHandCold** [comment]
@Mergifyio refresh

**Review by jiangxinmeng1 — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

**Comment by LeftHandCold** [comment]
@matrix-meow rerun failed

**Pull request merged**